src/OpenTelemetry.OpAmp.Client/FrameProcessor.cs

```diff
@@ -0,0 +1,123 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+using System.Buffers;
+using System.Collections.Concurrent;
+using System.Collections.Immutable;
+using OpAmp.Protocol;
+using OpenTelemetry.Internal;
+using OpenTelemetry.OpAmp.Client.Listeners;
+using OpenTelemetry.OpAmp.Client.Listeners.Messages;
+
+namespace OpenTelemetry.OpAmp.Client;
+
+internal class FrameProcessor
+{
+    private readonly ConcurrentDictionary<Type, ImmutableList<IOpAmpListener>> listeners = [];
+
+    public void Subscribe<T>(IOpAmpListener<T> listener)
+        where T : IOpAmpMessage
+    {
+        Guard.ThrowIfNull(listener, nameof(listener));
+
+        // It is expected to be much more read-heavy than write-heavy, so we use ImmutableList for thread safety
+        this.listeners.AddOrUpdate(
+            typeof(T),
+            _ => [listener],
+            (_, list) => list.Add(listener));
+    }
+
+    public void Unsubscribe<T>(IOpAmpListener<T> listener)
+        where T : IOpAmpMessage
+    {
+        Guard.ThrowIfNull(listener, nameof(listener));
+
+        this.listeners.AddOrUpdate(
+            typeof(T),
+            _ => ImmutableList<IOpAmpListener>.Empty,
+            (_, list) =>
+            {
+                if (list.Count == 1 && list[0] == listener)
+                {
+                    return ImmutableList<IOpAmpListener>.Empty;
+                }
+
+                return list.Remove(listener);
+            });
+    }
+
+    public void OnServerFrame(ReadOnlySequence<byte> sequence)
+    {
+        this.Deserialize(sequence);
+    }
+
+    private void Deserialize(ReadOnlySequence<byte> sequence)
+    {
+        var message = ServerToAgent.Parser.ParseFrom(sequence);
+
+        if (message.ErrorResponse != null)
```

Comment:
Would it be possible to collect all the messages to dispatch, and then send them together as a batch?

Reply:
Basically `ServerToAgent` is a "batch" in this case. It's deconstructed (fan-out) into smaller bits, so all interested parties can subscribe only to a necessary part. One `ServerToAgent` does not contain multiple same types.